Script to fetch and export huggingface LLMs

guangy10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #2331

Differential Revision: [D54733819](https://our.internmc.facebook.com/intern/diff/D54733819)

guangy10 committed:
Script to fetch and export huggingface LLMs

[ghstack-poisoned]

guangy10 commented:
@guangy10 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54733819).

dbort commented on examples/models/export_hf_model.py:
```diff
@@ -0,0 +1,178 @@
+                partitioners[
+                    XnnpackDynamicallyQuantizedPartitioner.__name__
+                ] = XnnpackDynamicallyQuantizedPartitioner()
+                prog = edge_m.to_backend(partitioners).to_executorch()
```
If you're having problems emitting large models, it should help to pass `ExecutorchBackendConfig(extract_constant_segment=True, extract_delegate_segments=True)` to `to_executorch()`. This will reduce the size of the JSON that is provided to `flatc`.

guangy10 replied:
@dbort W/ the suggested config emitting the gpt-neox-20b still crashes my devserver silently and I suspect if it's challenging the limit of flatc and would like to get more debug info. Would `ET_EXIR_SAVE_FLATC_INPUTS_ON_FAILURE =1` be helpful? And what is the right way to set this env var, something like `set ET_EXIR_SAVE_FLATC_INPUTS_ON_FAILURE=1 && python <my_script>`?

dbort replied:
> Would `ET_EXIR_SAVE_FLATC_INPUTS_ON_FAILURE =1` be helpful?

It could help debug the issue. It's an environment variable, so you can set it with `export ET_EXIR_SAVE_FLATC_INPUTS_ON_FAILURE=1` in your terminal, then running your export scripts.

dbort commented on examples/models/export_hf_model.py:
```diff
@@ -0,0 +1,178 @@
+                print(f"Lowered graph:\n{prog.exported_program().graph}")
+                filename = os.path.join(args.output_dir, f"xnnpack_{model_name}.pte")
+                with open(filename, "wb") as f:
+                    f.write(prog.buffer)
```
@lucylq just landed a new API at https://github.com/pytorch/executorch/pull/2307/files#diff-006db5ac7da9220ba2c903d189243689ccfd41d23f27d363702dbc133c8187ecR483 that can significantly reduce the memory consumed during emitting/saving. To use it here, you could say
```
prog.write_to_file(f)
```
This will avoid allocating a single buffer that contains the entire .pte data; instead, it writes individual buffers to the file directly.

mcr229 commented on examples/models/export_hf_model.py:
```diff
@@ -0,0 +1,178 @@
+
+from examples.portable.utils import export_to_edge
+from executorch.backends.xnnpack.partition.xnnpack_partitioner import (
+    XnnpackDynamicallyQuantizedPartitioner,
```
this partitioner should only be used with dynamically quantized models, if we're not using the quantization flow, then we should try with XnnpackPartitioner instead.

kimishpatel replied:
@mcr229 just add a patch to the PR?

guangy10 replied:
Switched to use the `XnnpackPartitioner` and rerun the lowering. Only 2 models previously succeeded fail with error
```
error:
  /tmp/tmpty09bomb/data.json:1: 128624: error: cannot parse value starting with: -
``` 
which happens during serialization, by then `XnnpackPartitioner` should have been run through successfully, correct? @mcr229

larryliu0820 commented on examples/models/export_hf_model.py:
```diff
@@ -0,0 +1,178 @@
+                tokenizer([prompt], return_tensors="pt").to(device)["input_ids"],
+            )
+
+            edge_m = export_to_edge(
```
Let's add a torch.no_grad() before this
```suggestion
            with torch.no_grad():
                edge_m = export_to_edge(
```

guangy10 committed:
Update on "Script to fetch and export huggingface LLMs"


Differential Revision: [D54733819](https://our.internmc.facebook.com/intern/diff/D54733819)

[ghstack-poisoned]

guangy10 committed:
Update on "Script to fetch and export huggingface LLMs"


Differential Revision: [D54733819](https://our.internmc.facebook.com/intern/diff/D54733819)

[ghstack-poisoned]
